File: replace_elements_with_greatest_element_on_right_side/replace_elements_with_greatest_element_on_right_side.java
// Prettier O(n) solution
class Solution {
    public int[] replaceElements(int[] arr) {
        int highestVal = arr[arr.length-1];
        for (int i = arr.length - 2; i >= 0; i--) {
            int tmpVal = arr[i];
            arr[i] = highestVal;
            if (tmpVal > highestVal) {
                highestVal = tmpVal;
            }
        }
        arr[arr.length-1] = -1;
        return arr;
    }
}

// O(n)2 solution with check from left
class Solution {
    public int[] replaceElements(int[] arr) {
        int highestVal;
        for (int i = 0; i < arr.length; i++) {
            if (i < arr.length - 1) {
                highestVal = arr[i+1];
                for (int j = i + 2; j < arr.length; j++) {
                    if (arr[j] > highestVal) {
                        highestVal = arr[j];
                    }
                }
                arr[i] = highestVal;
            
            } else {
                arr[i] = -1;
            }
        }
        return arr;
    }
}

File: replace_elements_with_greatest_element_on_right_side/replace_elements_with_greatest_element_on_right_side.py
class Solution:
    def replaceElements(self, arr: List[int]) -> List[int]:
        highest_val = arr[-1];
        for i in range(len(arr)-2, -1, -1):
            tmp_val = arr[i]
            arr[i] = highest_val
            if (tmp_val > highest_val):
                highest_val = tmp_val
            
        arr[-1] = -1
        return arr